### PR TITLE
Adds FourByteEncoder class to hold all the four byte encoding methods

### PR DESCRIPTION
**Pull request by LinZhihao-723**
# Description
Add a new Python class `FourByteEncoder` to hold all the four byte encoding methods as static members, and thus providing a better code organization.

# Validation performed
1. Passed all the existing unit tests.
2. Added new tests for FourByteEncoder class.
3. Passed both cpp and Python linting tools.

### CODE REVIEW

**Comment by LinZhihao-723** [comment]
The following will be the results of `help`:
```
class FourByteEncoder(builtins.object)
     |  Namespace for all CLP encoding methods.
     |  
     |  Methods encode bytes from the log record to create a CLP log message. This class should never be instantiated since it only contains static methods.
     |  
     |  Static methods defined here:
     |  
     |  encode_message(msg)
     |      Encodes the log `msg` using the 4-byte encoding.
     |      :param msg: Log message to encode.
     |      :raises NotImplementedError: If the log message failed to encode.
     |      :return: The encoded message.
     |  
     |  encode_message_and_timestamp_delta(timestamp_delta, msg)
     |      Encodes the log `msg` along with the timestamp delta using the 4-byte encoding.
     |      :param timestamp_delta: Timestamp difference in milliseconds between the current log message and the previous log message.
     |      :param msg: Log message to encode.
     |      :raises NotImplementedError: If the log message failed to encode, or the timestamp delta exceeds the supported size.
     |      :return: The encoded message and timestamp.
     |  
     |  encode_preamble(ref_timestamp, timestamp_format, timezone)
     |      Creates the encoded CLP preamble for a stream of encoded log messages using the 4-byte encoding.
     |      :param ref_timestamp: Reference timestamp used to calculate deltas emitted with each message.
     |      :param timestamp_format: Timestamp format to be use when generating the logs with a reader.
     |      :param timezone: Timezone in TZID format to be use when generating the timestamp from Unix epoch time.
     |      :raises NotImplementedError: If metadata length too large.
     |      :return: The encoded preamble.
     |  
     |  encode_timestamp_delta(timestamp_delta)
     |      Encodes the timestamp using the 4-byte encoding.
     |      :param timestamp_delta: Timestamp difference in milliseconds between the current log message and the previous log message.
     |      :raises NotImplementedError: If the timestamp failed to encode.
     |      :return: The encoded timestamp.
```

File: src/clp_ffi_py/modules/ir.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+PyModuleDef Py_ir{
+        PyModuleDef_HEAD_INIT,
+        "ir",
+        static_cast<char const*>(cModuleDoc),
+        -1,
+        static_cast<PyMethodDef*>(Py_ir_method_table)};
+}  // namespace
+
+extern "C" {
```
Comment (davidlion) [inline src/clp_ffi_py/modules/ir.cpp]:
```suggestion
```

File: src/clp_ffi_py/modules/ir.cpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    if (false == clp_ffi_py::ir::PyFourByteEncoder::module_level_init(new_module)) {
+        Py_DECREF(new_module);
+        return nullptr;
+    }
+
+    return new_module;
+}
+}
```
Comment (davidlion) [inline src/clp_ffi_py/modules/ir.cpp]:
```suggestion
```

Matches removal of opening `extern "C"` block on line 24.

File: tests/test_clpir.py [hunk 3]
```diff
@@ -276,5 +276,46 @@ def __check_log_event(
         self.assertEqual(idx, expected_idx, f"Message idx: {idx}, Expected: {expected_idx}")
 
 
+class TestCaseFourByteEncoder(unittest.TestCase):
+    """
+    Class for testing clp_ffi_py.CLPIR.FourByteEncoder.
```
Comment (davidlion) [inline tests/test_clpir.py]:
```suggestion
    Class for testing clp_ffi_py.ir.FourByteEncoder.
```

File: src/clp_ffi_py/ir/PyFourByteEncoder.hpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+#define CLP_FFI_PY_PyFourByteEncoder_HPP
+
+#include <clp_ffi_py/Python.hpp>  // Must always be included before any other header files
+
+#include <clp_ffi_py/PyObjectUtils.hpp>
+
+namespace clp_ffi_py::ir {
+class PyFourByteEncoder {
+    /**
+     * This class provides a Python level namespace for CLP 4-byte encoding methods.
```
Comment (davidlion) [inline src/clp_ffi_py/ir/PyFourByteEncoder.hpp]:
```suggestion
     * This class provides a Python level namespace for CLP 4-byte encoding
     * methods.
```

File: src/clp_ffi_py/ir/PyFourByteEncoder.hpp [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+     */
+public:
+    /**
+     * Creates and initializes PyFourByteEncoder as a Python type, and then
+     * incorporates this type as a Python object into the py_module module.
```
Comment (davidlion) [inline src/clp_ffi_py/ir/PyFourByteEncoder.hpp]:
```suggestion
     * incorporates this type as a Python object into the module, py_module.
```

File: src/clp_ffi_py/ir/PyFourByteEncoder.cpp [hunk 6]
```diff
@@ -1,14 +1,16 @@
+
```
Comment (davidlion) [inline src/clp_ffi_py/ir/PyFourByteEncoder.cpp]:
```suggestion
```

File: src/clp_ffi_py/ir/PyFourByteEncoder.cpp [hunk 7]
```diff
@@ -1,14 +1,16 @@
...
 #include <clp_ffi_py/Python.hpp>  // Must always be included before any other header files
 
-#include <clp_ffi_py/ir/encoding_methods.hpp>
-#include <clp_ffi_py/ir/PyLogEvent.hpp>
-#include <clp_ffi_py/ir/PyMetadata.hpp>
-#include <clp_ffi_py/Py_utils.hpp>
+#include "PyFourByteEncoder.hpp"
+
```
Comment (davidlion) [inline src/clp_ffi_py/ir/PyFourByteEncoder.cpp:4]:
```suggestion

#include <clp_ffi_py/ir/encoding_methods.hpp>
```

File: src/clp_ffi_py/ir/PyFourByteEncoder.cpp [hunk 8]
```diff
@@ -1,14 +1,16 @@
...
+#include <clp_ffi_py/PyObjectCast.hpp>
+#include <clp_ffi_py/utils.hpp>
 
-// CPython macros use C-arrays
-// NOLINTBEGIN(cppcoreguidelines-avoid-c-arrays)
+#include "encoding_methods.hpp"
```
Comment (davidlion) [inline src/clp_ffi_py/ir/PyFourByteEncoder.cpp]:
```suggestion
```

Added above without relative path.

File: src/clp_ffi_py/ir/PyFourByteEncoder.cpp [hunk 9]
```diff
@@ -59,60 +64,67 @@ PyDoc_STRVAR(
...
+         METH_VARARGS | METH_STATIC,
          static_cast<char const*>(cEncodeTimestampDeltaDoc)},
 
         {nullptr, nullptr, 0, nullptr}};
 
-PyDoc_STRVAR(cModuleDoc, "Python interface to the CLP IR encoding and decoding methods.");
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays)
+PyDoc_STRVAR(
+        cPyFourByteEncoderDoc,
+        "Namespace for all CLP encoding methods.\n\n"
```
Comment (davidlion) [inline src/clp_ffi_py/ir/PyFourByteEncoder.cpp]:
```suggestion
        "Namespace for all CLP four byte IR encoding methods.\n\n"
```

File: src/clp_ffi_py/ir/PyFourByteEncoder.cpp [hunk 10]
```diff
@@ -59,60 +64,67 @@ PyDoc_STRVAR(
...
+        "Methods encode bytes from the log record to create a CLP log message. This class should "
+        "never be instantiated since it only contains static methods.\n"
+);
 
-struct PyModuleDef PyCLPIR {
-    PyModuleDef_HEAD_INIT, "CLPIR", static_cast<char const*>(cModuleDoc), -1,
-            static_cast<PyMethodDef*>(PyCLPIR_method_table)
-};
+// NOLINTBEGIN(cppcoreguidelines-avoid-c-arrays, cppcoreguidelines-pro-type-*-cast)
```
Comment (davidlion) [inline src/clp_ffi_py/ir/PyFourByteEncoder.cpp]:
```suggestion
// NOLINTBEGIN(cppcoreguidelines-avoid-c-arrays, cppcoreguidelines-pro-type-const-cast)
```

File: src/clp_ffi_py/ir/PyFourByteEncoder.cpp [hunk 11]
```diff
@@ -59,60 +64,67 @@ PyDoc_STRVAR(
...
+PyType_Slot PyFourByteEncoder_slots[]{
+        {Py_tp_methods, static_cast<void*>(PyFourByteEncoder_method_table)},
+        {Py_tp_doc, const_cast<void*>(static_cast<void const*>(cPyFourByteEncoderDoc))},
+        {0, nullptr}};
+// NOLINTEND(cppcoreguidelines-avoid-c-arrays, cppcoreguidelines-pro-type-*-cast)
```
Comment (davidlion) [inline src/clp_ffi_py/ir/PyFourByteEncoder.cpp]:
```suggestion
// NOLINTEND(cppcoreguidelines-avoid-c-arrays, cppcoreguidelines-pro-type-const-cast)
```

File: tests/test_clpir.py [hunk 12]
```diff
@@ -276,5 +276,46 @@ def __check_log_event(
...
+        encoder.
+
+        It only checks if the result of
+        encode_message_and_timestamp_delta is consistent with the combine of
+        encode_message and encode_timestamp_delta. The actual functionality
+        should be covered by the unittest of CLP Python logging library.
+        TODO: When the decoder is implemented, add some more tests to ensure
+        the encoded bytes can be successfully decoded to recover the original
+        log event.
+        """
```
Comment (davidlion) [inline tests/test_clpir.py]: I think we can just rename this test (and rewrite the doc string accordingly) to be explicitly for testing whether "`encode_message_and_timestamp_delta` is consistent with the addition of `encode_message` and `encode_timestamp_delta`".

We can move the TODO and comments on general encoding tests to the class doc string.

**Review by davidlion — CHANGES_REQUESTED** [review]

File: tests/test_clpir.py [hunk 13]
```diff
@@ -292,18 +297,11 @@ def test_init(self) -> None:
...
-
-        It only checks if the result of
-        encode_message_and_timestamp_delta is consistent with the combine of
-        encode_message and encode_timestamp_delta. The actual functionality
-        should be covered by the unittest of CLP Python logging library.
-        TODO: When the decoder is implemented, add some more tests to ensure
-        the encoded bytes can be successfully decoded to recover the original
-        log event.
+        This test checks if the result of encode_message_and_timestamp_delta is
+        consistent with the combine of encode_message and
```
Comment (davidlion) [inline tests/test_clpir.py]:
```suggestion
        consistent with the conbination of encode_message and
```

File: src/clp_ffi_py/ir/PyFourByteEncoder.hpp [hunk 14]
```diff
@@ -8,12 +8,13 @@
...
     /**
-     * This class provides a Python level namespace for CLP 4-byte encoding methods.
+     * This class provides a Python level namespace for CLP 4-byte encoding
+     * methods.
      */
 public:
     /**
      * Creates and initializes PyFourByteEncoder as a Python type, and then
-     * incorporates this type as a Python object into the py_module module.
+     * incorporates this type as a Python object into the py_module.
```
Comment (davidlion) [inline src/clp_ffi_py/ir/PyFourByteEncoder.hpp]:
```suggestion
     * incorporates this type as a Python object into py_module.
```

**Review by davidlion — CHANGES_REQUESTED** [review]

File: src/clp_ffi_py/ir/PyFourByteEncoder.hpp [hunk 15]
```diff
@@ -0,0 +1,32 @@
+#ifndef CLP_FFI_PY_PyFourByteEncoder_HPP
+#define CLP_FFI_PY_PyFourByteEncoder_HPP
```
Comment (kirkrodrigues) [inline src/clp_ffi_py/ir/PyFourByteEncoder.hpp]:
```suggestion
#ifndef CLP_FFI_PY_PYFOURBYTEENCODER_HPP
#define CLP_FFI_PY_PYFOURBYTEENCODER_HPP
```

File: src/clp_ffi_py/ir/PyFourByteEncoder.hpp [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+#include <clp_ffi_py/Python.hpp>  // Must always be included before any other header files
+
+#include <clp_ffi_py/PyObjectUtils.hpp>
+
+namespace clp_ffi_py::ir {
+class PyFourByteEncoder {
+    /**
+     * This class provides a Python level namespace for CLP 4-byte encoding
+     * methods.
+     */
```
Comment (kirkrodrigues) [inline src/clp_ffi_py/ir/PyFourByteEncoder.hpp]:
```suggestion
/**
 * This class provides a Python-level namespace for CLP 4-byte IR encoding
 * methods.
 */
class PyFourByteEncoder {
```

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]
* Minor fixes.
* You still have an instance of "clp_ir.cpp" in `clp_ffi_py/ir/encoding_methods.hpp`.

**Review by kirkrodrigues — APPROVED** [review]

**Review by davidlion — APPROVED** [review]

**Pull request merged**